snapd-glib/snapd-client.c

```diff
@@ -2443,16 +2462,22 @@ snapd_client_connect_sync (SnapdClient *client,
  */
 void
 snapd_client_connect_async (SnapdClient *client,
-                            GCancellable *cancellable, GAsyncReadyCallback callback, gpointer user_data)
+                            GCancellable *cancellable,
+                            GAsyncReadyCallback callback,
+                            gpointer user_data)
 {
-    GTask *task;
-    g_autoptr(GError) error = NULL;
+    SnapdClientPrivate *priv;
+    g_autoptr(GTask) task = NULL;
+
+    g_return_if_fail (SNAPD_IS_CLIENT (client));
+
+    priv = snapd_client_get_instance_private (client);
 
+    g_main_context_push_thread_default (priv->context);
     task = g_task_new (client, cancellable, callback, user_data);
-    if (snapd_client_connect_sync (client, cancellable, &error))
-        g_task_return_boolean (task, TRUE);
-    else
-        g_task_return_error (task, error);
+    g_main_context_pop_thread_default (priv->context);
+
+    g_task_run_in_thread (task, do_connect_thread);
```

Comment:
I don't *think* there's a problem in the old way. If you want, I'll revert this bit and we can maybe look at it later. There's a bit of an overhead in doing it all sync, so it might be worth thinking about, but I can see that it's probably not necessary for this PR.